### PR TITLE
Update Yaml Configuration support

### PR DESCRIPTION
**Pull request by buddhikac96**
**Updates**
- Add sample-allOption.yml (There are few differences in config names compared to XML configuration to keep UX)
- Update time span values to parsing from strings. Now user can specify TimeSpan values in `5 sec` manner.
- Implement unimplemented properties in YamlConfiguration.cs
- Remove unnecessary properties from ServiceDescriptorYaml.cs

**Possible Updates**
- Document the configurations is sample-allOption.yml file

 **On-going updates**
- WinSW YAML Support Documentation.

### CODE REVIEW

File: src/Core/WinSWCore/Configuration/YamlConfiguration.cs [hunk 1]
```diff
@@ -496,19 +513,44 @@ public void LoadEnvironmentVariables()
...
+        public string BaseName => this.Defaults.BaseName;
 
-        public string BasePath => throw new NotImplementedException();
+        public string BasePath => this.Defaults.BasePath;
 
-        public string? ServiceAccountDomain => throw new NotImplementedException();
+        public string? SecurityDescriptor => this.SecurityDescriptorYaml ?? this.Defaults.SecurityDescriptor;
 
-        public string? ServiceAccountName => throw new NotImplementedException();
+        public static TimeSpan ParseTimeSpan(string v)
```
Comment (oleg-nenashev) [inline src/Core/WinSWCore/Configuration/YamlConfiguration.cs:527]: Code duplication from XML? I would suggest to move it to a separate utility class then

File: src/Core/WinSWCore/Configuration/YamlConfiguration.cs [hunk 2]
```diff
@@ -100,6 +100,9 @@ public class YamlConfiguration : IWinSWConfiguration
         [YamlMember(Alias = "securityDescriptor")]
         public string? SecurityDescriptorYaml { get; set; }
 
+        [YamlMember(Alias = "extensions")]
+        public List<string>? YamlExtensionIds { get; set; }
```
Comment (oleg-nenashev) [inline src/Core/WinSWCore/Configuration/YamlConfiguration.cs:104]: This one will not work as defined, right? We still need to support extensions somehow

File: src/Core/WinSWCore/Configuration/YamlConfiguration.cs (hunk 1 above)
Comment (buddhikac96) [reply src/Core/WinSWCore/Configuration/YamlConfiguration.cs:527]: Make sense. will update.

File: src/Core/WinSWCore/Configuration/YamlConfiguration.cs (hunk 2 above)
Comment (buddhikac96) [reply src/Core/WinSWCore/Configuration/YamlConfiguration.cs:104]: Yes. It is my next plan on YAML Support

File: examples/sample-allOption.yml [hunk 3]
```diff
@@ -0,0 +1,72 @@
+id: myapp
+name: MyApp Service (powered by WinSW)
+description: This service is a service created from a sample configuration
+executable: java
+caption: WinSW
```
Comment (NextTurn) [inline examples/sample-allOption.yml:5]: There is no corresponding entry in XML.

File: examples/sample-allOption.yml [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+#serviceaccount:
+#    domain: YOURDOMAIN
+#    user: useraccount
+#    password: Pa55w0rd
+#    allowservicelogon: yes
+#failureActions:
```
Comment (NextTurn) [inline examples/sample-allOption.yml:11]: I don't think `failureActions` is better than `onFailure`.

File: examples/sample-allOption.yml [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+#securityDescriptor: security descriptor string
+#arguments: >
+#    -classpath 
+#    c:\cygwin\home\kohsuke\ws\hello-world\out\production\hello-world 
+#    test.Main
+#startArguments: start arguments
+#workingdirectory: C:\myApp\work
+priority: Normal
+stopTimeout: 15 sec
+stopParentProcessFirst: false
```
Comment (NextTurn) [inline examples/sample-allOption.yml:30]: Use `true` in samples.

File: examples/sample-allOption.yml (hunk 4 above)
Comment (buddhikac96) [reply examples/sample-allOption.yml:11]: will change into `onFailure`

File: examples/sample-allOption.yml (hunk 3 above)
Comment (buddhikac96) [reply examples/sample-allOption.yml:5]: Its a mistake. Should be removed.

File: examples/sample-allOption.yml [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+#    - Eventlog
+#    - W32Time
+waitHint: 15 sec
+sleepTime: 1 sec
+#interactive: true
+log:
+#    logpath: '%BASE%\logs'
+    mode: append
+#env:
+#    MY_TOOL_HOME: 'C:\etc\tools\myTool'
```
Comment (NextTurn) [inline examples/sample-allOption.yml:44]: - Shall we support a similar syntax for XML in v3?
- Shall we support `name:` and `value:` pairs?

Comment (buddhikac96) [reply examples/sample-allOption.yml:44]: Sure. I am okay with that! I will update this.

Comment (buddhikac96) [reply examples/sample-allOption.yml:44]: > * Shall we support a similar syntax for XML in v3

Sorry, I did not get this???

Comment (oleg-nenashev) [reply examples/sample-allOption.yml:44]: I believe @NextTurn asks whether we should change XML syntax in v3 so that YAML and XML are aligned

Comment (buddhikac96) [reply examples/sample-allOption.yml:44]: What would be good?? I am open for both. However I already change YAML config into XML syntax. I will push that. But still I can roll back to previous syntax if you wish.

File: doc/YamlConfigFile.md [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+YAML Configuration file shuold be in following format
+
+Example:
+
+```yaml
+id: jenkins
+name: Jenkins
+description: This service runs Jenkins continuous integration system.
+env:
+    -
```
Comment (NextTurn) [inline doc/YamlConfigFile.md:19]: Blank lines aren't required.

File: src/Core/WinSWCore/Configuration/YamlConfiguration.cs [hunk 8]
```diff
@@ -315,15 +325,29 @@ public class YamlDownload
...
+            public SC_ACTION_TYPE Type
+            {
+                get
+                {
+                    SC_ACTION_TYPE actionType = this.FailureAction switch
+                    {
+                        "restart" => SC_ACTION_TYPE.SC_ACTION_RESTART,
+                        "none" => SC_ACTION_TYPE.SC_ACTION_NONE,
+                        "reboot" => SC_ACTION_TYPE.SC_ACTION_REBOOT,
+                        _ => throw new Exception("Invalid failure action: " + this.FailureAction)
```
Comment (NextTurn) [inline src/Core/WinSWCore/Configuration/YamlConfiguration.cs:343]: Don't throw `Exception`.

File: doc/YamlConfigFile.md [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+```
+
+See the [Logging and error reporting](loggingAndErrorReporting.md) page for more info.
+
+### Arguments
+
+`arguments` element specifies the arguments to be passed to the executable. User can specify all the commands as a single line.
+
+```yaml
+args: arg1 arg2 arg3
```
Comment (NextTurn) [inline doc/YamlConfigFile.md:107]: As per the discussion in Gitter we don't want abbreviations. Can you confirm @oleg-nenashev?

File: src/Core/WinSWCore/Configuration/YamlConfiguration.cs (hunk 8 above)
Comment (buddhikac96) [reply src/Core/WinSWCore/Configuration/YamlConfiguration.cs:343]: Which value should be bound to actionType for a invalid input?

Comment (NextTurn) [reply src/Core/WinSWCore/Configuration/YamlConfiguration.cs:343]: Throw `InvalidDataException` ~~or `InvalidEnumArgumentException`~~ instead.

File: doc/YamlConfigFile.md (hunk 9 above)
Comment (buddhikac96) [reply doc/YamlConfigFile.md]: Updated

File: doc/YamlConfigFile.md [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+    -jar "%BASE%\jenkins.war"
+    --httpPort=8080
+log:
+    mode: roll
+```
+
+## Environment variable expansion
+
+Configuration YAML files can include environment variable expansions of the form `%Name%`. 
+Such occurrences, if found, will be automatically replaced by the actual values of the variables. [Read more. . . ](xmlConfigFile.md#environment-variable-expansion)
```
Comment (oleg-nenashev) [inline doc/YamlConfigFile.md:34]: Is it really supported at the moment? I do not see it in code though I am looking from phone and might be missing it

File: doc/YamlConfigFile.md [hunk 11]
```diff
@@ -0,0 +1,298 @@
...
+
+## Configuration entries
+
+### id
```
Comment (oleg-nenashev) [inline doc/YamlConfigFile.md:38]: There is a lot of duplication between this doc and XML. Maybe we should move detailed docs into a separate doc and link them.

It does not block this PR, can be done later. Docs need update for v3 anyway

**Comment by buddhikac96** [comment]
> Looks good overall. Just a few comments. I'd guess.we want to proceed without the extensions implementation here, right?

Yeah. At the moment I am working on that on another branch.

File: doc/YamlConfigFile.md (hunk 11 above)
Comment (buddhikac96) [reply doc/YamlConfigFile.md:38]: Indeed. I am also thinking its more user friendly. Anyway if we move into `winsw.github.io` then we can present that in much more user friendly way.

File: doc/YamlConfigFile.md (hunk 10 above)
Comment (buddhikac96) [reply doc/YamlConfigFile.md:34]: I guess we are talking about [this](https://github.com/buddhikac96/winsw/blob/9dd7fb4d93055a582628ce8d963b1016d9715c38/src/Core/WinSWCore/Configuration/YamlConfiguration.cs#L500)??

Comment (oleg-nenashev) [reply doc/YamlConfigFile.md:34]: Nope. Syntax like `  <executable>%BASE%\myExecutable.exe</executable>` in XML when the variable is actually used in other fields. This is what `Environment variable expansion` means

File: doc/YamlConfigFile.md [hunk 12]
```diff
@@ -0,0 +1,298 @@
+# YAML configuration file
+
+This page describes the configuration file, which controls the behavior of the Windows service.
+
+You can find configuration file samples in the [examples](../examples) directory of the source code repository.
+Actual samples are being also published as a part of releases on GitHub and NuGet.
```
Comment (NextTurn) [inline doc/YamlConfigFile.md:6]:
```suggestion
Actual samples are also being published as a part of releases on GitHub and NuGet.
```

File: doc/YamlConfigFile.md (hunk 10 above)
Comment (NextTurn) [inline doc/YamlConfigFile.md:34]: `Read more. . . ` looks weird.

File: doc/YamlConfigFile.md [hunk 13]
```diff
@@ -0,0 +1,298 @@
+# YAML configuration file
+
+This page describes the configuration file, which controls the behavior of the Windows service.
+
+You can find configuration file samples in the [examples](../examples) directory of the source code repository.
+Actual samples are also being published as a part of releases on GitHub and NuGet.
```
Comment (NextTurn) [inline doc/YamlConfigFile.md:6]:
```suggestion
Actual samples are also being published as part of releases on GitHub and NuGet.
```

/cc @oleg-nenashev

File: doc/YamlConfigFile.md (hunk 10 above)
Comment (buddhikac96) [reply doc/YamlConfigFile.md]: Updated

File: doc/YamlConfigFile.md (hunk 13 above)
Comment (oleg-nenashev) [reply doc/YamlConfigFile.md]: I am also not a native speaker :( 
I think we will be doing a major cleanup for v3 anyway

**Pull request merged**